tools/state-viewer/src/commands.rs

```diff
@@ -1161,49 +1150,131 @@ pub(crate) fn print_state_stats(home_dir: &Path, store: Store, near_config: Near
+    block_hash: CryptoHash,
+    shard_uid: ShardUId,
+) {
+    flat_storage_manager.create_flat_storage_for_shard(shard_uid).unwrap();
+    let trie_storage = TrieDBStorage::new(store.clone(), shard_uid);
+    let chunk_view = flat_storage_manager.chunk_view(shard_uid, block_hash).unwrap();
 
-        let iter = chunk_view.iter_flat_state_entries(None, None);
-        for item in iter {
-            let Ok((key, value)) = item else {
-                continue;
-            };
+    let mut state_stats = StateStats::default();
 
-            let value = read_flat_state_value(&trie_storage, value);
+    // iteratate for the first time to get the size statistics
+    let group_by = get_state_stats_group_by(&chunk_view, &trie_storage);
+    let iter = get_state_stats_account_iter(&group_by);
+    for state_stats_account in iter {
+        state_stats.push(state_stats_account);
+    }
 
-            state_stats.push(key, value);
+    // iterate for the second time to find the middle account
+    let group_by = get_state_stats_group_by(&chunk_view, &trie_storage);
+    let iter = get_state_stats_account_iter(&group_by);
+    let mut current_size = ByteSize::default();
+    for state_stats_account in iter {
+        current_size += state_stats_account.size;
+        if 2 * current_size.as_u64() > state_stats.total_size.as_u64() {
+            state_stats.middle_state_record = Some(state_stats_account);
+            break;
         }
+    }
 
-        let middle_size = state_stats.total_size.as_u64() / 2;
-
-        let mut current_size = 0;
-        let iter = chunk_view.iter_flat_state_entries(None, None);
-        for item in iter {
-            let Ok((key, value)) = item else {
-                tracing::warn!(target: "state_viewer", ?item, "error in iter item");
-                continue;
-            };
+    tracing::info!(target: "state_viewer", "{shard_uid:?}");
+    tracing::info!(target: "state_viewer", "{state_stats:#?}");
+}
 
-            let value = read_flat_state_value(&trie_storage, value);
+#[derive(Eq, PartialEq)]
+struct StateStatsStateRecord {
+    account_id: AccountId,
+    state_record: StateRecord,
+    size: ByteSize,
+}
 
-            current_size += key.len() + value.len();
-            if middle_size <= current_size as u64 {
-                state_stats.middle_state_record = Some(StateStatsStateRecord { key, value });
-                break;
-            }
-        }
+impl Ord for StateStatsStateRecord {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.account_id.cmp(&other.account_id)
+    }
+}
 
-        tracing::info!(target: "state_viewer",  "{shard_uid:?}");
-        tracing::info!(target: "state_viewer",  "{state_stats:#?}");
+impl PartialOrd for StateStatsStateRecord {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
     }
 }
 
+// Gets the flat state iterator and groups items by account id.
+fn get_state_stats_group_by<'a>(
```

Comment:
Aah this is very clever! Took me some time to understand what's going on, had to read up on the kmerge and group_by functions.

Reply:
It took me real long to get the types, lifespans and whatnots in order but yeah it's pretty now :)